pkg/controller/elasticsearch/client/error.go

```diff
@@ -0,0 +1,92 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License 2.0;
+// you may not use this file except in compliance with the Elastic License 2.0.
+
+package client
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+)
+
+// APIError is a non 2xx response from the Elasticsearch API
+type APIError struct {
+	Status        string
+	StatusCode    int
+	ErrorResponse ErrorResponse
+}
+
+// newAPIError converts an HTTP response into an API error, attempting to parse the body to include the details about the error.
+func newAPIError(response *http.Response) error {
+	defer response.Body.Close()
+	apiError := &APIError{
+		Status:     response.Status,
+		StatusCode: response.StatusCode,
+	}
+	// We may need to read the body multiple times, read the full body and store it as an array of bytes.
+	body, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		// We were not able to read the body, log this I/O error and return the API error with the status.
+		log.Error(err, "Cannot read error body")
+		return apiError
+	}
+	// Reset the response body to the original unread state. It allows a caller to read again the body if necessary,
+	// for example in the case of a 408.
+	response.Body = ioutil.NopCloser(bytes.NewBuffer(body))
+
+	// Parse the body to get the details about the API error, as they are stored by Elasticsearch.
+	var errorResponse ErrorResponse
+	if err := json.Unmarshal(body, &errorResponse); err != nil {
+		// Only log at the debug level since it is expected to not be able to parse all types of errors.
+		// Some errors, like 408 on /_cluster/health may return a different body structure.
+		log.V(1).Error(err, "Cannot parse error body")
+		return apiError
+	}
+	apiError.ErrorResponse = errorResponse
+	return apiError
+}
+
+// Error() implements the error interface.
```

Comment:
```suggestion
// APIError implements the error interface.
```

Reply:
I _think_ the purpose of this comment was to mention that `Error()` allows `APIError` to implement `error`. I would rather fix it as follow to respect godoc convention:

```suggestion
// Error implements the error interface.
```

Reply:
++ to what you said :)